mesh/exodus_writer.cpp

```diff
@@ -0,0 +1,1018 @@
+
+   const FiniteElementSpace * fespace = _mesh.GetNodalFESpace();
+
+   mfem::Array<int> element_dofs;
+   for (int ielement = 0; ielement < _mesh.GetNE(); ielement++)
+   {
+      if (fespace)   // Higher-order
+      {
+         fespace->GetElementDofs(ielement, element_dofs);
+      }
+      else
+      {
+         _mesh.GetElementVertices(ielement, element_dofs);
+      }
+
+      for (int dof : element_dofs)
+      {
+         unique_node_ids.insert(dof);
+      }
+   }
+
+   return unique_node_ids;
+}
+
+void ExodusIIWriter::GenerateExodusIIBoundaryInfo()
+{
+   // Store the unique boundary IDs.
+   _boundary_ids.clear();
+   _exodusII_element_ids_for_boundary_id.clear();
+   _exodusII_side_ids_for_boundary_id.clear();
+
+   for (int bdr_attribute : _mesh.bdr_attributes)
+   {
+      _boundary_ids.push_back(bdr_attribute);
+   }
+
+   // Generate a mapping from the MFEM face index to the MFEM element ID.
+   // Note that if we have multiple element IDs for a face index then the
+   // face is shared between them and it cannot possibly be a boundary face
+   // since that can only have a single element associated with it. Therefore
+   // we remove it from the array.
+   struct GlobalFaceIndexInfo
+   {
+      int element_index;
+      int local_face_index;
+
+      GlobalFaceIndexInfo() : element_index{0}, local_face_index{0} {}
+
+      GlobalFaceIndexInfo(int element_index, int local_face_index)
+      {
+         this->element_index = element_index;
+         this->local_face_index = local_face_index;
+      }
+   };
+
+   std::unordered_map<int, GlobalFaceIndexInfo>
+   mfem_face_index_info_for_global_face_index;
+   std::unordered_set<int> blacklisted_global_face_indices;
+
+   Array<int> global_face_indices, orient;
+   for (int ielement = 0; ielement < _mesh.GetNE(); ielement++)
+   {
+      _mesh.GetElementFaces(ielement, global_face_indices, orient);
+
+      for (int iface = 0; iface < global_face_indices.Size(); iface++)
+      {
+         int face_index = global_face_indices[iface];
+
+         if (blacklisted_global_face_indices.count(face_index))
+         {
+            continue;
+         }
+
+         if (mfem_face_index_info_for_global_face_index.count(face_index))
+         {
+            // Now we've seen it twice!
+            blacklisted_global_face_indices.insert(face_index);
+            mfem_face_index_info_for_global_face_index.erase(face_index);
+            continue;
+         }
+
+         mfem_face_index_info_for_global_face_index[face_index] = GlobalFaceIndexInfo(
+                                                                     ielement, iface);
+      }
+   }
+
+   for (int ibdr_element = 0; ibdr_element < _mesh.GetNBE(); ibdr_element++)
+   {
+      int boundary_id = _mesh.GetBdrAttribute(ibdr_element);
+      int bdr_element_face_index = _mesh.GetBdrElementFaceIndex(ibdr_element);
+
+      // Locate match.
+      auto & element_face_info = mfem_face_index_info_for_global_face_index.at(
```

Comment:
I'm not certain but I believe this call to `at` may be failing on interior surfaces such as those found in `miniapps/multidomain/multidomain-hex.mesh`. Some would argue that "boudary" faces should never exist within the interior of a mesh but MFEM meshes do support this. Unfortunately this means that this writer must determine how best to handle such things. Does the Exodus II format support interior "boundary" faces? Does it store such things in a different manner? At the very least the current implementation of this writer should recognize these "interior" boundary faces and skip them if necessary (and issue a warning message to that effect).

Reply:
Right, thanks for spotting this. I've had a quick scan through the ExodusII file format specification and it is not entirely clear. The best solution for now would be to identify the interior boundary faces and skip them. I'll look into doing this.

Reply:
We could either call _RemoveInternalBoundaries_ to remove these or probably better is to check the boundary by calling _FaceIsInterior_. This would avoid modifying the mesh we are writing to the ExodusII file.

Reply:
Okay, I've pushed a commit which addresses the issue with miniapps/multidomain/multidomain-hex.mesh. It calls FaceIsInterior to check whether a boundary element is internal and if it is then we print a warning and skip it.